cpp/oneapi/dal/backend/primitives/ndarray.hpp

```diff
@@ -0,0 +1,405 @@
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include "oneapi/dal/array.hpp"
+#include "oneapi/dal/backend/memory.hpp"
+#include "oneapi/dal/backend/primitives/ndshape.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+enum class ndorder {
+    c, /* C-style ordering, row-major in 2D case */
+    f /* Fortran-style ordering, column-major in 2D case */
+};
+
+template <ndorder order>
+struct transposed_ndorder;
+
+template <>
+struct transposed_ndorder<ndorder::c> {
+    static constexpr ndorder value = ndorder::f;
+};
+
+template <>
+struct transposed_ndorder<ndorder::f> {
+    static constexpr ndorder value = ndorder::c;
+};
+
+template <ndorder order>
+constexpr ndorder transposed_ndorder_v = transposed_ndorder<order>::value;
+
+template <std::int64_t axis_count, ndorder order = ndorder::c>
+class ndarray_base {
+    static_assert(axis_count > 0, "Axis count must be non-zero");
+    static_assert(order == ndorder::c || order == ndorder::f, "Only C or F orders are supported");
+
+public:
+    ndarray_base() = default;
+
+    explicit ndarray_base(const ndshape<axis_count>& shape, const ndshape<axis_count>& strides)
+            : shape_(shape),
+              strides_(strides) {
+        if constexpr (order == ndorder::c) {
+            ONEDAL_ASSERT(strides[axis_count - 1] == 1, "Last C-order stride must be 1");
+        }
+        else if constexpr (order == ndorder::f) {
+            ONEDAL_ASSERT(strides[0] == 1, "First F-order stride must be 1");
+        }
+
+#ifdef ONEDAL_ENABLE_ASSERT
+        const auto default_strides = get_default_strides(shape);
+        for (std::int64_t i = 0; i < axis_count; i++) {
+            ONEDAL_ASSERT(strides[i] >= default_strides[i],
+                          "Custom stride must be greater than default");
+        }
+#endif
+    }
+
+    explicit ndarray_base(const ndshape<axis_count>& shape)
+            : ndarray_base(shape, get_default_strides(shape)) {}
+
+    ndorder get_order() const {
+        return order;
+    }
+
+    const ndshape<axis_count>& get_shape() const {
+        return shape_;
+    }
+
+    const ndshape<axis_count>& get_strides() const {
+        return strides_;
+    }
+
+    std::int64_t get_shape(std::int64_t axis) const {
+        return shape_[axis];
+    }
+
+    std::int64_t get_stride(std::int64_t axis) const {
+        return strides_[axis];
+    }
+
+    std::int64_t get_leading_stride() const {
+        if constexpr (order == ndorder::c) {
+            return strides_[0];
+        }
+        else if constexpr (order == ndorder::f) {
+            return strides_[axis_count - 1];
+        }
+        return 0;
+    }
+
+    std::int64_t get_count() const {
```

Comment:
Why strides do not affect the count of elements?

Reply:
You think they should? This method returns number of elements in the matrix, not size of the allocated memory.